### ISSUE TITLE
vscode-remote locks ssh

### ISSUE THREAD

**Issue by niallel** [microsoft/vscode-remote-release#305]
Issue Type: <b>Bug</b>

After a while of using with an amazon linux EC2 instance the connection drops and doesn't reconnect.
Also at this point I can't ssh into the server using the os terminal.
I have to reboot the server to get access again.

The logs I see on the server when I manage to connect again are:
```
[2019-05-13 17:25:07.099] [exthost] [info] extension host started
[2019-05-13 17:25:07.340] [exthost] [info] ExtensionService#_doActivateExtension vscode.typescript-language-features {"startup":false,"activationEvent":"onLanguage:javascript"}
[2019-05-13 17:25:07.341] [exthost] [info] ExtensionService#loadCommonJSModule /home/ec2-user/.vscode-remote/bin/57b550c559b945eb9d871dbf2b2e4cb9e31e2551/extensions/typescript-language-features/dist/extension
[2019-05-13 17:25:07.384] [exthost] [info] ExtensionService#_doActivateExtension vscode.debug-auto-launch {"startup":true,"activationEvent":"*"}
[2019-05-13 17:25:07.384] [exthost] [info] ExtensionService#loadCommonJSModule /home/ec2-user/.vscode-remote/bin/57b550c559b945eb9d871dbf2b2e4cb9e31e2551/extensions/debug-auto-launch/dist/extension
[2019-05-13 17:25:07.388] [exthost] [info] ExtensionService#_doActivateExtension vscode.emmet {"startup":true,"activationEvent":"*"}
[2019-05-13 17:25:07.388] [exthost] [info] ExtensionService#loadCommonJSModule /home/ec2-user/.vscode-remote/bin/57b550c559b945eb9d871dbf2b2e4cb9e31e2551/extensions/emmet/dist/extension
[2019-05-13 17:25:07.397] [exthost] [info] ExtensionService#_doActivateExtension vscode.git {"startup":true,"activationEvent":"*"}
[2019-05-13 17:25:07.398] [exthost] [info] ExtensionService#loadCommonJSModule /home/ec2-user/.vscode-remote/bin/57b550c559b945eb9d871dbf2b2e4cb9e31e2551/extensions/git/dist/main
[2019-05-13 17:25:07.467] [exthost] [info] ExtensionService#_doActivateExtension vscode.merge-conflict {"startup":true,"activationEvent":"*"}
[2019-05-13 17:25:07.467] [exthost] [info] ExtensionService#loadCommonJSModule /home/ec2-user/.vscode-remote/bin/57b550c559b945eb9d871dbf2b2e4cb9e31e2551/extensions/merge-conflict/dist/extension
[2019-05-13 17:25:07.473] [exthost] [info] eager extensions activated
[2019-05-13 17:26:48.799] [exthost] [info] ExtensionService#_doActivateExtension vscode.extension-editing {"startup":false,"activationEvent":"onLanguage:typescript"}
[2019-05-13 17:26:48.800] [exthost] [info] ExtensionService#loadCommonJSModule /home/ec2-user/.vscode-remote/bin/57b550c559b945eb9d871dbf2b2e4cb9e31e2551/extensions/extension-editing/dist/extension
[2019-05-13 17:41:43.276] [exthost] [error] [vscode.typescript-language-features] provider FAILED
[2019-05-13 17:41:43.278] [exthost] [error] Error: TypeScript Server Error (3.4.5)
Debug Failure. Invalid cast. The supplied value [object Object] did not pass the test 'isLabeledStatement'.
Error: Debug Failure. Invalid cast. The supplied value [object Object] did not pass the test 'isLabeledStatement'.
    at Object.cast (tsserver.js:1434:22)
    at doChange (tsserver.js:115512:39)
    at tsserver.js:115504:96
    at Function.ChangeTracker.with (tsserver.js:110610:17)
    at Object.getCodeActions (tsserver.js:115504:64)
    at tsserver.js:111596:121
    at Object.flatMap (tsserver.js:571:25)
    at Object.getFixes (tsserver.js:111596:23)
    at tsserver.js:121245:35
    at Object.flatMap (tsserver.js:571:25)
    at Object.getCodeFixesAtPosition (tsserver.js:121243:23)
    at IOSession.Session.getCodeFixes (tsserver.js:130292:64)
    at Session.handlers.ts.createMapFromTemplate._a.(anonymous function) (tsserver.js:129048:61)
    at tsserver.js:130466:88
    at IOSession.Session.executeWithRequestId (tsserver.js:130457:28)
    at IOSession.Session.executeCommand (tsserver.js:130466:33)
    at IOSession.Session.onMessage (tsserver.js:130488:35)
    at Interface.<anonymous> (tsserver.js:131789:27)
    at Interface.emit (events.js:182:13)
    at Interface._onLine (readline.js:290:10)
    at Interface._normalWrite (readline.js:433:12)
    at Socket.ondata (readline.js:149:10)
    at Socket.emit (events.js:182:13)
    at addChunk (_stream_readable.js:279:12)
    at readableAddChunk (_stream_readable.js:264:11)
    at Socket.Readable.push (_stream_readable.js:219:10)
    at Pipe.onread (net.js:636:20)
	at Function.create (/home/ec2-user/.vscode-remote/bin/57b550c559b945eb9d871dbf2b2e4cb9e31e2551/extensions/typescript-language-features/dist/extension.js:1:176236)
	at v.dispatchResponse (/home/ec2-user/.vscode-remote/bin/57b550c559b945eb9d871dbf2b2e4cb9e31e2551/extensions/typescript-language-features/dist/extension.js:1:181604)
	at v.dispatchMessage (/home/ec2-user/.vscode-remote/bin/57b550c559b945eb9d871dbf2b2e4cb9e31e2551/extensions/typescript-language-features/dist/extension.js:1:180356)
	at _reader.onData.e (/home/ec2-user/.vscode-remote/bin/57b550c559b945eb9d871dbf2b2e4cb9e31e2551/extensions/typescript-language-features/dist/extension.js:1:179712)
	at u.fire (/home/ec2-user/.vscode-remote/bin/57b550c559b945eb9d871dbf2b2e4cb9e31e2551/out/vs/agent/remoteExtensionHostProcess.js:47:720)
	at t.Reader.onLengthData (/home/ec2-user/.vscode-remote/bin/57b550c559b945eb9d871dbf2b2e4cb9e31e2551/extensions/typescript-language-features/dist/extension.js:1:185585)
	at Socket.e.on.e (/home/ec2-user/.vscode-remote/bin/57b550c559b945eb9d871dbf2b2e4cb9e31e2551/extensions/typescript-language-features/dist/extension.js:1:185233)
	at Socket.emit (events.js:182:13)
	at addChunk (_stream_readable.js:279:12)
	at readableAddChunk (_stream_readable.js:264:11)
	at Socket.Readable.push (_stream_readable.js:219:10)
	at Pipe.onread (net.js:636:20)
```

VS Code version: Code - Insiders 1.34.20-insider (57b550c559b945eb9d871dbf2b2e4cb9e31e2551, 2019-05-10T17:35:18.558Z)
OS version: Darwin x64 18.5.0
Fetching remote diagnostics for 'SSH' failed.

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-6820HQ CPU @ 2.70GHz (8 x 2700)|
|GPU Status|2d_canvas: enabled<br>checker_imaging: disabled_off<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>native_gpu_memory_buffers: enabled<br>rasterization: enabled<br>surface_synchronization: enabled_on<br>video_decode: enabled<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|1, 1, 2|
|Memory (System)|16.00GB (0.07GB free)|
|Process Argv||
|Screen Reader|no|
|VM|15%|

Fetching remote diagnostics for 'SSH' failed.
</details><details><summary>Extensions (6)</summary>

Extension|Author (truncated)|Version
---|---|---
remote-containers|ms-|0.53.0
remote-ssh|ms-|0.37.0
remote-ssh-edit|ms-|0.37.0
remote-ssh-explorer|ms-|0.37.0
remote-wsl|ms-|0.34.0
vscode-remote-extensionpack|ms-|0.14.0

</details>

**Comment by mjbvz** [comment]
Moving as the ssh bug is likely not related to tsserver (and even if it is, an extension should not be able to stop reconnection)

**Comment by roblourens** [comment]
> Also at this point I can't ssh into the server using the os terminal.

Can you explain some more? What are the logs if you add `-vvv` to your ssh command?

**Comment by niallel** [comment]
I've just tried doing it again and I realised that my whole server is uncontactable.

It responds to ping, but none of the other services are responding.

Is there some debugging I can turn on, on the server to help?

**Comment by roblourens** [comment]
I don't know what the problem would be. What info does amazon give you? Does it show the cpu or disk going crazy?

**Comment by niallel** [comment]
It shows the CPU going up sharply before the failure.

Interestingly, I created a brand new amazon image and played about with it for an hour, it all worked fine. Then I cloned my git repo to the host, and within 10 minutes the problem happened.

If it helps I can give you access?

Happy to help in anyway I can, I think this vscode-remote thing is amazing, it does everything I've wanted for years.

**Comment by roblourens** [comment]
Sure I could try your rep.

Can you tell me what the specs on your amazon vm are?

Did you run/debug your app remotely, or just develop? Which extensions do you have installed on the remote machine?

If you keep experimenting it would help to open a terminal, run `top`, and see which process uses lots of CPU before the failure.

**Comment by niallel** [comment]
I tried to create a machine for you, but without my repo and using the linux kernel instead.
At the moment its not crashing, however git only shows the changes in vscode if I quit and restart. Any changes I make are not shown. I suppose this is a different problem.

I did the following:
- Created Amazon Linux AMI 2018.03.0 (HVM), SSD Volume Type - ami-08935252a36e25f85
- T2.Micro size, with 100GB disk
- Added port 80 to the security groups
- Updated the packages with sudo yum update
- Installed git and apache with sudo yum install git httpd
- Rebooted the server with sudo reboot
- I then created an entry in my ssh config file with 
Host 18.202.19.12
  HostName 18.202.19.12
  User ec2-user
  IdentityFile ~/vscode-test.pem

- Next I cloned the linux kernel with git clone https://github.com/torvalds/linux.git

Lastly I connected with visual studio code and started messing with some files, making changes saving them etc.

I'm going to recreate this machine with my git repo and see if I can make it happen again.

Thanks for all the help so far.
